spring-messaging/src/main/java/org/springframework/messaging/handler/annotation/support/PayloadMethodArgumentResolver.java

```diff
@@ -151,9 +147,9 @@ public Object resolveArgument(MethodParameter parameter, Message<?> message) thr
 				throw new MessageConversionException(message, "Cannot convert from [" +
 						payloadClass.getName() + "] to [" + targetClass.getName() + "] for " + message);
 			}
-			validate(message, parameter, payload);
-			return (isOptionalTargetClass ? Optional.of(payload) : payload);
 		}
+		validate(message, parameter, payload);
+		return (isOptionalTargetClass ? Optional.of(payload) : payload);
```

Comment:
I wonder if it'd be more natural to early return when ClassUtils.isAssignable, then the rest don't have to be nested (not sure why it originally nested).

Reply:
But then you would be back to having duplicate logic since validation and `Optional` wrapping is required regardless. Also, conversion is likely the more common case anyway, so it's okay for that to be performed first.